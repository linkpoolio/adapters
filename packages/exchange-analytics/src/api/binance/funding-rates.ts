import { util } from '@chainlink/ea-bootstrap'

import FundingRate from '../../models/funding-rate'
import type { IFundingRate } from '../../models/funding-rate'
import { Provider } from '../constants'

export default (fetch) => ({
  get: async ({ asset }): Promise<IFundingRate> => {
    const url = util.buildUrlPath(`premiumIndex?symbol=:assetUSDT`, {
      asset,
    })
    const response = await fetch({ url })
    return FundingRate.Single(response.data, Provider.BINANCE)
  },
})
