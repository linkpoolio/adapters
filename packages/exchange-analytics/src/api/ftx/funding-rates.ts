import { util } from '@chainlink/ea-bootstrap'

import FundingRate from '../../models/funding-rate'
import type { IFundingRate } from '../../models/funding-rate'
import { Provider } from '../constants'

export default (fetch) => ({
  get: async ({ asset }): Promise<IFundingRate> => {
    const url = util.buildUrlPath(`futures/:asset-PERP/stats`, {
      asset,
    })
    const response = await fetch({ url })
    return FundingRate.Single(response.data, Provider.FTX)
  },
})
