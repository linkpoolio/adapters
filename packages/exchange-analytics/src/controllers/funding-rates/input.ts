import type { InputParameters } from '@chainlink/types'

import { Asset } from '../constants'
import { sharedInputParameters } from '../input'

export const getSingleInputParameters: InputParameters = {
  ...sharedInputParameters,
  asset: {
    description: 'The name of the asset.',
    required: true,
    type: 'string',
    options: [Asset.ETH, Asset.BTC],
  },
}

export const inputParameters: InputParameters = {
  ...getSingleInputParameters,
}
