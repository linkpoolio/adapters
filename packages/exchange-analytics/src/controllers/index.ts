export * as fudingRates from './funding-rates'
