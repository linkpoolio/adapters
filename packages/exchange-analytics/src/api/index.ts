import { AdapterError, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

import binance from './binance'
import bybit from './bybit'
import { Provider } from './constants'
import ftx from './ftx'

export default (config: Config) => {
  const provider = util.getRequiredEnv('API_PROVIDER')

  switch (provider) {
    case Provider.BINANCE:
      return binance(config)
    case Provider.BYBIT:
      return bybit(config)
    case Provider.FTX:
      return ftx(config)
    default: {
      const message = `Unsupported provider: ${provider}. Check API_PROVIDER env var`
      throw new AdapterError({
        statusCode: 500,
        message,
        cause: message,
      })
    }
  }
}
