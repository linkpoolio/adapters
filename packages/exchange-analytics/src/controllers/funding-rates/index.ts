import execute from './controller'
import { inputParameters } from './input'

const supportedEndpoints = ['funding-rates']
const endpointResultPaths = {}
const description = 'This endpoint returns the funding rates for ETH or BTC paired with USD.'

export { supportedEndpoints, endpointResultPaths, description, inputParameters, execute }
