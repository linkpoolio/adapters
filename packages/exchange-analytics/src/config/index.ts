import { AdapterError, Requester, util } from '@chainlink/ea-bootstrap'
import { enums } from '@linkpool/shared'

import { Provider } from '../api/constants'
import type { ApiProviderConfig, SupportedApiProviderConfig } from './types'

export const NAME = 'EXCHANGE-ANALYTICS'

export const makeConfig = (prefix?: string): SupportedApiProviderConfig => {
  const config: ApiProviderConfig = {
    ...Requester.getDefaultConfig(prefix),
    apiProvider: util.getRequiredEnv('API_PROVIDER'),
  }

  switch (config.apiProvider) {
    case Provider.BINANCE: {
      return {
        ...config,
      }
    }
    case Provider.BYBIT: {
      return {
        ...config,
      }
    }
    case Provider.FTX: {
      return {
        ...config,
      }
    }
    default: {
      const message = `Unsupported provider: ${
        config.apiProvider
      }. Supported providers are: ${enums.formatEnumValuesPretty(
        Provider as unknown as Record<string, number>,
      )}. Check API_PROVIDER env var`
      throw new AdapterError({
        statusCode: 500,
        message,
        cause: message,
      })
    }
  }
}
