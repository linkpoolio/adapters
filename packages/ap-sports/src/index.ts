import { expose } from '@chainlink/ea-bootstrap'

import { endpointSelector, makeExecute } from './adapter'
import { NAME, makeConfig } from './config'
import * as rateLimit from './config/limits.json'
import * as endpoints from './endpoint'
import type * as types from './lib/types'

const adapterContext = { name: NAME, rateLimit }

const { server } = expose(adapterContext, makeExecute(), undefined, endpointSelector)
export { NAME, makeExecute, makeConfig, server, types, endpoints }
