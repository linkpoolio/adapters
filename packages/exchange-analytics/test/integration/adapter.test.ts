import http from 'http'
import type { AddressInfo } from 'net'
import process from 'process'
import request from 'supertest'
import type { SuperTest, Test } from 'supertest'

import { server as startServer } from '../../src'
import { testFundingRates } from './funding-rates/get'

let oldEnv: NodeJS.ProcessEnv

export interface SuiteContext {
  server: http.Server
  req: SuperTest<Test>
}

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.CACHE_ENABLED = 'false'
  process.env.WARMUP_ENABLED = 'false'
  process.env.API_KEY = 'fake-api-key'
})

afterAll(() => {
  process.env = oldEnv
})

describe('execute', () => {
  const context: SuiteContext = {
    server: null,
    req: null,
  }

  beforeEach(async () => {
    context.server = await startServer()
    context.req = request(`localhost:${(context.server.address() as AddressInfo).port}`)
  })

  afterEach((done) => {
    context.server.close(done)
  })

  describe('exchange-analytics', () => {
    describe('funding-rates endpoint', () => testFundingRates(context))
  })
})
