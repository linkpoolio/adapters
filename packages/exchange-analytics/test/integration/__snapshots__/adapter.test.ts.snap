// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute endpoint funding-rates endpoint binance returns a single funding rate 1`] = `
Object {
  "data": Object {
    "nextFunding": Object {
      "rate": -0.000019615725081026844,
      "timestamp": 1663344000,
    },
  },
  "jobRunID": 1,
  "statusCode": 200,
}
`;

exports[`execute endpoint funding-rates endpoint binance returns an error response when the API errors too 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"funding-rates\\",\\"method\\":\\"get\\",\\"asset\\":\\"ETH\\"}}",
    "message": "Unexpected error requesting the 'funding-rates' endpoint. Reason: AdapterError: Request failed with status code 500",
    "name": "AdapterError",
    "url": "https:/fapi.binance.com/fapi/v1/premiumIndex?symbol=ETHUSDT",
  },
  "jobRunID": 1,
  "providerStatusCode": 500,
  "status": "errored",
  "statusCode": 200,
}
`;

exports[`execute endpoint funding-rates endpoint bybit returns a single funding rate 1`] = `
Object {
  "data": Object {
    "nextFunding": Object {
      "rate": -0.00001119125,
      "timestamp": 1663315200,
    },
  },
  "jobRunID": 1,
  "statusCode": 200,
}
`;

exports[`execute endpoint funding-rates endpoint bybit returns an error response when the API errors too 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"funding-rates\\",\\"method\\":\\"get\\",\\"asset\\":\\"ETH\\"}}",
    "message": "Unexpected error requesting the 'funding-rates' endpoint. Reason: AdapterError: Request failed with status code 500",
    "name": "AdapterError",
    "url": "https:/api.bybit.com/v2/public/funding/prev-funding-rate?symbol=ETHUSD",
  },
  "jobRunID": 1,
  "providerStatusCode": 500,
  "status": "errored",
  "statusCode": 200,
}
`;

exports[`execute endpoint funding-rates endpoint ftx returns a single funding rate 1`] = `
Object {
  "data": Object {
    "nextFunding": Object {
      "rate": -0.000005,
      "timestamp": 1663336800,
    },
  },
  "jobRunID": 1,
  "statusCode": 200,
}
`;

exports[`execute endpoint funding-rates endpoint ftx returns an error response when the API errors too 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"funding-rates\\",\\"method\\":\\"get\\",\\"asset\\":\\"ETH\\"}}",
    "message": "Unexpected error requesting the 'funding-rates' endpoint. Reason: AdapterError: Request failed with status code 500",
    "name": "AdapterError",
    "url": "https:/ftx.com/api/futures/ETH-PERP/stats",
  },
  "jobRunID": 1,
  "providerStatusCode": 500,
  "status": "errored",
  "statusCode": 200,
}
`;
